server-spi/src/main/java/org/keycloak/models/GroupProvider.java

```diff
@@ -0,0 +1,181 @@
+package org.keycloak.models;
+
+import org.keycloak.provider.Provider;
+
+import java.util.List;
+
+/**
+ *
+ * Provider of group records
+ * @author mhajas
+ *
+ */
+public interface GroupProvider extends Provider {
+
+    /**
+     * Returns a group from the given realm with the corresponding id
+     * 
+     * @param id Id.
+     * @param realm Realm.
+     * @return GroupModel with the corresponding id.
+     * @deprecated Use method {@code getGroupById(realm, id)}
+     */
+    default GroupModel getGroupById(String id, RealmModel realm) {
+        return getGroupById(realm, id);
+    }
+
+    /**
+     * Returns a group from the given realm with the corresponding id
+     *
+     * @param realm Realm.
+     * @param id Id.
+     * @return GroupModel with the corresponding id.
+     */
+    GroupModel getGroupById(RealmModel realm, String id);
+
+    /**
+     * Returns groups for the given realm.
+     *
+     * @param realm Realm.
+     * @return List of groups in the Realm.
+     */
+    List<GroupModel> getGroups(RealmModel realm);
+
+    /**
+     * Returns a number of groups/top level groups (i.e. groups without parent group) for the given realm.
+     *
+     * @param realm Realm.
+     * @param onlyTopGroups When true the function returns a count of top level groups only.
+     * @return Number of groups/top level groups.
+     */
+    Long getGroupsCount(RealmModel realm, Boolean onlyTopGroups);
+
+    /**
+     * Returns number of groups with the given string in name for the given realm.
+     *
+     * @param realm Realm.
+     * @param search Searched string.
+     * @return Number of groups with the given string in its name.
+     */
+    Long getGroupsCountByNameContaining(RealmModel realm, String search);
+
+    /**
+     * Returns groups with the given role in the given realm.
+     *
+     * @param realm Realm.
+     * @param role Role.
+     * @param firstResult First result to return. Ignored if negative.
+     * @param maxResults Maximum number of results to return. Ignored if negative.
+     * @return List of groups with the given role.
+     */
+    List<GroupModel> getGroupsByRole(RealmModel realm, RoleModel role, int firstResult, int maxResults);
```

Comment:
Shouldn't we have Stream variants for these methods?

Reply:
This is going to be part of https://github.com/keycloak/keycloak/pull/7328